core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java

```diff
@@ -236,6 +245,24 @@ public void flush() throws IOException {
                   mAddress, mPartialRequest, WRITE_TIMEOUT_MS));
         }
       }
+      pos = mPosToQueue;
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new CanceledException(e);
+    }
+    if (mEOFSent || mCancelSent) {
+      return;
+    }
+    Protocol.WriteRequest writeRequest =
+        mPartialRequest.toBuilder().setOffset(pos).setFlush(true).build();
+    mChannel.writeAndFlush(new RPCProtoMessage(new ProtoMessage(writeRequest), null))
+        .addListener(new FlushListener());
+    try (LockResource lr = new LockResource(mLock)) {
+      if (!mFlushedOrDoneOrFailed.await(FLUSH_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+        throw new DeadlineExceededException(
+            String.format("Timeout flush to %s for request %s after %dms.",
+                mAddress, mPartialRequest, FLUSH_TIMEOUT_MS));
+      }
```

Comment:
I think the code from `if (mEOFSent || mCancelSent)` to `mChannel.writeAndFlush` also needs to be guarded by `mLock`. Otherwise, an `EOF` or `Cancel` can be sent after the check but before the flush. This causes the `FLUSH` packet to be behind an `EOF` or `Cancel` packet which terminates the processing on the server side, resulting in the flush hanging.

Reply:
Oh yes! mEOFSent and mCancelSent are also guarded by the mLock.
So now, when the flush is running, none thread is able to send EOF or Cancel.
Great!!! :clap: Thanks for helping debugging this!

Reply:
Right, we won't need to hold the lock until flush completes, just until we send the flush packet, since as long as the flush is before EOF or Cancel, it will be processed.